File: 144_interleaving-positive-and-negative-numbers/interleaving-positive-and-negative-numbers.cpp
/*
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/interleaving-positive-and-negative-numbers
@Language: C++
@Datetime: 15-05-29 01:30
*/

class Solution {
public:
    /**
     * @param A: An integer array.
     * @return an integer array
     */
    vector<int> rerange(vector<int> &A) {
        int i = 0, j = 0, n = A.size();
        for ( ; j < n; ++j) 
            if (A[j] < 0 )  swap(A[i++], A[j]);
        if(i > n - i) reverse(A.begin(), A.end());
        for (i = (i > n - i ? 0 : 1), j = 1 ^ i; i < n && j < n;  ) {
            while ( i < n && A[i] < 0) i += 2;
            while ( j < n && A[j] > 0) j += 2;
            if (i < n && j < n) swap (A[i], A[j]);
        }
        return A;
    }
};

File: 144_interleaving-positive-and-negative-numbers/interleaving-positive-and-negative-numbers.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   heiyanbin
@Problem:  http://www.lintcode.com/problem/interleaving-positive-and-negative-numbers
@Language: Python
@Datetime: 14-12-23 10:46
'''

class Solution:
    """
    @param A: An integer array.
    @return an integer array
    """
    def rerange(self, A):
        i, j, n =0, 0, len(A)
        for j in xrange (0, n):
            if A[j] < 0:
                A[i], A[j] = A[j], A[i]
                i += 1
        i, j = (0, 1) if i >= n - i else (1, 0)
        while i<n and j<n:
            while i<n and A[i]<0: i += 2
            while j<n and A[j]>0: j += 2
            if i < n and j < n: A[i], A[j] = A[j], A[i]
        return A
              
